### PR TITLE
 Allow the catchall pattern inside of nested patterns

### LINKED ISSUE #1449: Allow the catchall pattern inside of nested patterns
Right now "scrutinees" can only hold other "scrutinees", and cannot hold the catchall pattern

### PR DESCRIPTION
**Pull request by emilyaherbert**
This is built off of #1592 

Allows you to create patterns like:

```rust
let e = (
    (1, 2),
    (
        (3, 4),
        5
    )
);
let f = match e {
    ((3, _), _) => { 99 },
    (_, (_, 5)) => { 42 },
    _ => { 0 },
};
```

Closes  #1449

### CODE REVIEW

File: test/src/e2e_vm_tests/test_programs/should_pass/language/match_expressions_simple/src/main.sw [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        (
+            (3, 4),
+            5
+        )
+    );
+    let f = match e {
+        ((3, _), _) => { 99 },
+        (_, (_, 5)) => { 42 },
+        _ => { 0 },
+    };
```
Comment (mitchmindtree) [inline test/src/e2e_vm_tests/test_programs/should_pass/language/match_expressions_simple/src/main.sw:30]: So cool, feels like we're getting close to Rust-level pattern matching!

File: sway-core/src/semantic_analysis/ast_node/expression/usefulness/mod.rs [hunk 2]
```diff
@@ -202,47 +202,31 @@ use self::{
...
+        let v = PatStack::from_pattern(pat);
+        let witness_report = check!(
+            is_useful(&factory, &matrix, &v, &span),
+            return err(warnings, errors),
+            warnings,
+            errors
+        );
+        matrix.push(v);
+        // if an arm has witnesses to its usefulness then it is reachable
+        arms_reachability.push((scrutinee.clone(), witness_report.has_witnesses()));
```
Comment (mitchmindtree) [inline sway-core/src/semantic_analysis/ast_node/expression/usefulness/mod.rs:229]: So much cleaner, nice!

**Review by mitchmindtree — APPROVED** [review]
Such a pleasant read! Adds super useful functionality while cutting the code into something much more concise :heart_eyes: 

I left a comment about the test that's up to you to handle, happy to re-approve if you want to address it.

Otherwise utACK LGTM.

File: test/src/e2e_vm_tests/test_programs/should_pass/language/match_expressions_structs/src/main.sw [hunk 3]
```diff
@@ -5,15 +5,33 @@ struct Point {
...
+    };
+    let d = match c {
+        Data { value: false } => { 0 },
+        Data { value } => { 4 },
+    };
+    // // should fail
+    // let e = match c {
+    //     Data { value: 1u64 } => { false },
+    //     Data { value } => { true },
+    // };
```
[the remark this answers is not in the document]
Comment (emilyaherbert) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/match_expressions_structs/src/main.sw:34]: Good call!

> should this give an unreachable branch error or something along those lines?

Yep! An unreachable compiler warning.

File: test/src/e2e_vm_tests/test_programs/should_pass/language/match_expressions_simple/src/main.sw (hunk 1 above)
Comment (emilyaherbert) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/match_expressions_simple/src/main.sw:30]: 😄

**Review by mitchmindtree — APPROVED** [review]

**Pull request merged**